enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,322 @@
+is omitted from this proposal.
+
+* HAProxy provides a performance tuning variable, `maxcomprate`, to set the maximum per-process input compression rate
+  to the specified number of kilobytes per second.  It works in conjunction with `tune.comp.maxlevel` to rate-limit the
+  compression level to save data bandwidth. Another, `maxcompcpuusage`, does the same to save limit CPU used on compression.
+Neither variable is currently exposed, and both are omitted from this proposal.  (Other per-session performance tuning
+variables like `maxconnrate`, which set the maximum per-process number of connections per second, and `maxsslrate`,
+which sets the maximum per-process number of SSL sessions per second, are not exposed either.)
+
+## Proposal
+
+Currently, the haproxy.cfg contains the lines below in the `defaults` section, but the mechanism for setting the
+environment variables is not exposed.
+```go
+defaults
+...
+{{- if isTrue (env "ROUTER_ENABLE_COMPRESSION") }}
+  compression algo gzip
+  compression type {{ env "ROUTER_COMPRESSION_MIME" "text/html text/plain text/css" }}
+  {{- end }}
+```
+
+### Global HTTP Compression via IngressControllerSpec and environment variables
+
+The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
+```go
+type IngressControllerSpec struct {
+...
+    // httpCompression defines policy for HTTP traffic compression
+    // 
+    // The default value is no compression
+    //
+    // +optional
+    HTTPCompression *HTTPCompressionPolicy `json:"httpCompression,omitempty"`
+}
+...
+// httpCompressionPolicy turns on compression for the specified MIME types
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the specified
+// MIME types
+type HTTPCompressionPolicy struct {
+	// compressionMIMETypes is a list of MIME types that should have compression applied.
+	// At least one MIME type must be specified.
+	//
+	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
+	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
+	// formats benefit from compression, but formats that are already compressed (pdf, image,
+	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
+	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	CompressionMIMETypes []CompressionMIMEType `json:"compressionMIMETypes,omitempty"`
+}
+
+// CompressionMIMEType defines the format of a single MIME type
+// E.g. "text/css; charset=utf-8", "text/html", "text/*", "image/svg+xml",
+// "application/octet-stream", etc.
+//
+// +kubebuilder:validation:Pattern="[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"
```

Comment:
This should be anchored.  You shouldn't need to escape `+` in a character class.  Are spaces really allowed in the type?  https://datatracker.ietf.org/doc/html/rfc1341#page-7 indicates that spaces, `;`, and `=` should be prohibited in the type and subtype.

Reply:
"text/css; charset=utf-8" is a valid MIME type, so I included spaces, `;`, and `=` in the pattern.

Reply:
And will add `"`.

Reply:
I kept this pretty loose.  RFC1341 is made obsolete by https://datatracker.ietf.org/doc/html/rfc1521#page-68  

I will anchor and remove escapes from `+`, but if we don't keep this loose, we need to use https://datatracker.ietf.org/doc/html/rfc1521#page-68